### PR TITLE
Field-dependent boundary conditions

### LINKED ISSUE #897: "Field-dependent" boundary condition functions
We have the need for continuous (not discrete) boundary condition functions that depend on the field to which the boundary condition is applied. For example, in the `ocean_convection_and_wind_mixing.jl`, the prescribed salinity flux depends on the salinity adjacent to the boundary:

https://github.com/CliMA/Oceananigans.jl/blob/f13a0a26df78cd1c1495d56ef5e7ea2fbbbfb23c/examples/ocean_wind_mixing_and_convection.jl#L73

The linear drag boundary condition implemented in `eady_turbulence.jl` is similar:

https://github.com/CliMA/Oceananigans.jl/blob/f13a0a26df78cd1c1495d56ef5e7ea2fbbbfb23c/examples/eady_turbulence.jl#L60

A slightly simpler interface would allow users to implement the salinity flux (for example) with

```julia
Qˢ(x, y, t, S, evaporation) = - evaporation * S
```

For this to work, a number of changes must be made in `BoundaryFunction`, `BoundaryCondition`, `FieldBoundaryConditions`, and the `TracerFields` constructor:

- `BoundaryFunction` must record whether or not the function is `field_dependent`, as well as the "`side`" of the boundary condition (left or right, so that the wall-normal index may be inferred), the `field_name`, and the `field_container` in which the field lives (`model.velocities` or `model.tracers`).

- `BoundaryCondition` must provide the option `field_dependent=true`.

- `FieldBoundaryConditions` must supply the `side` at which the boundary condition is applied (`Left()` for west, south, bottom and `Right()` for east, north, and top)

- `TracerFields` must supply the tracer name.

New methods for field-dependent `BoundaryFunction`s are also required. For `z`-boundaries, for example, we need

```julia
@inline function (bc::BoundaryFunction{:z, X, Y, true})(j, k, grid, clock, state) where {X, Y}
    k = wall_adjacent_index(bc.side, grid.Nz)
    container = getproperty(state, bc.field_container)
    field = getproperty(container, bc.field_name)
    
    return @inbounds call_boundary_function(bc.func, xnode(X, i, grid), ynode(Y, j, grid), clock.time, field[i, j, k], bc.parameters)
end
```

### PR DESCRIPTION
**Pull request by glwagner**
This PR implements a field-dependent boundary conditions feature. This is an extension to the boundary conditions API that leaves existing functionality unchanged.

The functionality is similar to that for `Forcing`. For quadratic bottom stress acting on the `u`-velocity we'd write

```julia
u_bottom_stress(x, y, t, u, v, Cd) = - Cd * u * sqrt(u^2 + v^2)

bottom_bc = BoundaryCondition(Flux, u_bottom_stress, field_dependencies=(:u, :v), parameters=1e-3)
```

Dealing with the wall-normal velocity components is a bit annoying. For now we don't interpolate them, which leads to different behavior at left and right boundaries (due to the fact that the wall-normal index is assumed to refer to cell centers). Right now this is not much of a practical issue --- it would only affect cases in which wall-normal velocities were non-zero _and_ those values were used in boundary conditions on other fields.

One solution is to store "wall-normal" indices separately for each of the `field_dependencies`. Then we can correctly infer the boundary index in `regularize_boundary_condition` because we know field locations there.

We still need to:

- [x] update boundary conditions docs
- [x] see if any examples can be simplified?

Resolves #897

### CODE REVIEW

File: test/utils_for_runtests.jl [hunk 1]
```diff
@@ -99,3 +99,27 @@ function run_script(replace_strings, script_name, script_filepath, module_suffix
...
+#####
+
+discrete_func(i, j, grid, clock, model_fields) = - model_fields.u[i, j, grid.Nz]
+parameterized_discrete_func(i, j, grid, clock, model_fields, p) = - p.μ * model_fields.u[i, j, grid.Nz]
+
+parameterized_fun(ξ, η, t, p) = p.μ * cos(p.ω * t)
+field_dependent_fun(ξ, η, t, u, v, w) = - w * sqrt(u^2 + v^2) 
+exploding_fun(ξ, η, t, T, S, p) = - p.μ * cosh(S - p.S0) * exp((T - p.T0) / p.λ)
+
+# Many bc. Very many
```
Comment (ali-ramadhan) [inline test/utils_for_runtests.jl:114]: Very important BCs

**Review by ali-ramadhan — APPROVED** [review]

File: examples/ocean_wind_mixing_and_convection.jl [hunk 2]
```diff
@@ -3,93 +3,116 @@
...
-         Δz = 1.0      # Grid spacing in x, y, z (meters)
-         Qᵀ = 5e-5     # Temperature flux at surface
-         Qᵘ = -2e-5    # Velocity flux at surface
-       ∂T∂z = 0.005    # Initial vertical temperature gradient
-evaporation = 1e-7     # Mass-specific evaporation rate [m s⁻¹]
-          f = 1e-4     # Coriolis parameter
-          α = 2e-4     # Thermal expansion coefficient
-          β = 8e-4     # Haline contraction coefficient
-nothing # hide
+# With surface temperature flux `Qʰ`, density `ρ`, and heat capacity `cᴾ`,
```
[the remark this answers is not in the document]
Comment (glwagner) [reply examples/ocean_wind_mixing_and_convection.jl:41]: I'm going to leave this change out, since these lines are updated in PR #1099

File: src/BoundaryConditions/continuous_boundary_function.jl [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        return new{X, Y, Z, I,
+                   typeof(func),
+                   typeof(parameters),
+                   typeof(field_dependencies),
+                   typeof(field_dependencies_indices),
+                   typeof(field_dependencies_interp)}(func,
+                                                      parameters,
+                                                      field_dependencies,
+                                                      field_dependencies_indices,
+                                                      field_dependencies_interp)
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/BoundaryConditions/continuous_boundary_function.jl:44]: would be better if this code created a pyramid

File: src/Operators/interpolation_utils.jl [hunk 4]
```diff
@@ -58,3 +58,30 @@ Return the `identity` interpolator function. This is needed to obtain the interp
 opertator for fields that have no instrinsic location, like numbers or functions.
 """
 interpolation_operator(::Nothing, to) = identity
+
+assumed_field_location(name) = name === :u ? (Face, Cell, Cell) :
+                               name === :v ? (Cell, Face, Cell) :
+                               name === :w ? (Cell, Cell, Face) :
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/Operators/interpolation_utils.jl:64]: There's no difference for symbols:

https://github.com/JuliaLang/julia/blob/bac321d949452784c946e9cc6cf1c05b14fc31a8/base/operators.jl#L55-L83

(there used to be the fallback `==(x, y) = x === y` but that definition lives somewhere else now...)

Ah, certainly we only need one definition #technical-debt

File: test/utils_for_runtests.jl (hunk 1 above)
Comment (glwagner) [reply test/utils_for_runtests.jl:114]: very many important bcs

File: src/Operators/interpolation_utils.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (glwagner) [reply src/Operators/interpolation_utils.jl:64]: Ok, there's only one definition now.
